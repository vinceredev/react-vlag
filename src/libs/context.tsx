import React, { createContext } from 'react';
import { useCookies } from 'react-cookie';

interface VlagPropsType {
  id: string;
  name: string;
  description?: string;
  helperText?: string;
  disabled: boolean;
}

type VlagProps = {
  children: React.ReactNode;
  flags: Array<VlagPropsType>;
};

interface VlagContextType {
  isEnabled: (id: string) => boolean;
  setEnable: (id: string, val: boolean) => void;
  getFlags: () => Array<VlagPropsType>;
  flagMap: Array<string>;
}

const VlagContext = createContext<VlagContextType>({
  isEnabled: () => false,
  setEnable: () => undefined,
  getFlags: () => [],
  flagMap: []
});

export const VlagProvider = ({ children, flags }: VlagProps) => {
  const flagMap = flags.map((fl) => fl.id);
  const [cookies, setCookie] = useCookies(flagMap);

  const isEnabled = (id: string) => {
    return cookies[id] === 'true';
  };

  const setEnable = (id: string, val: boolean) => {
    setCookie(id, val);
  };

  const getFlags = () => {
    return flags;
  };

  let hasDuplicate = flagMap.some((val, i) => flagMap.indexOf(val) !== i);
  if (hasDuplicate) {
    return <>flag id is duplicate</>;
  }

  return (
    <VlagContext.Provider value={{ isEnabled, setEnable, getFlags, flagMap }}>
      {children}
    </VlagContext.Provider>
  );
};

export const useVlag = () => React.useContext(VlagContext);
