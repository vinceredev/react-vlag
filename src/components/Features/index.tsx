import React from "react";
import { useVlag } from "../../libs/context";
import Switch from "../switch";

const Features: React.FC = () => {
  const vlag = useVlag();
  const vlagFeatures = vlag.getFlags();

  const vlagID = vlag.flagMap.reduce(
    (o, key) => Object.assign(o, { [key]: vlag.isEnabled(key) }),
    {}
  );
  const [enableFeature, setEnableFeature] = React.useState<any>(vlagID);

  const handleVlagChange = (key: string, val: boolean) => {
    setEnableFeature({ ...enableFeature, [key]: val });
    vlag.setEnable(key, val);
  };

  return (
    <div className="vlag-features">
      {vlagFeatures?.map((feature, i) => (
        <div key={i} className="vlag-features__item">
          <div className="vlag-features__item__content">
            <h2 className="vlag-features__item__content__title">
              {feature?.name}
            </h2>
            {feature?.description && (
              <p className="vlag-features__item__content__description">
                {feature?.description}
              </p>
            )}
            {feature?.helperText && (
              <p className="vlag-features__item__content__helper">
                {feature?.helperText}
              </p>
            )}
          </div>
          <div className="vlag-features__item__toggle">
            <Switch
              checked={enableFeature[feature?.id]}
              disabled={feature.disabled}
              onClick={() =>
                !feature.disabled && handleVlagChange(feature?.id, !enableFeature[feature?.id])
              }
            />
          </div>
        </div>
      ))}
    </div>
  );
};

export default Features;
