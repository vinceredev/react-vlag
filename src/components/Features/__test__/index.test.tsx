import React from 'react';
import renderer from 'react-test-renderer';
import Features from '../index';
import { VlagProvider } from '../../../libs/context';

it('success: vlag features renders correctly', () => {
  const vlags = (
    <VlagProvider
      flags={[
        {
          id: 'new_checkout_page',
          name: 'Checkout Page',
          description:
            'Lorem Ipsum is simply dummy text of the printing and typesetting industry. ',
          helperText: 'Will be available for beta testing end of 2022',
          disabled: false
        },
        {
          id: 'new_product_page',
          name: 'Product Page',
          description:
            'Lorem Ipsum is simply dummy text of the printing and typesetting industry. ',
          disabled: false
        }
      ]}
    >
      <Features />
    </VlagProvider>
  );

  const trees = renderer.create(vlags);
  expect(trees).toMatchSnapshot();
});
