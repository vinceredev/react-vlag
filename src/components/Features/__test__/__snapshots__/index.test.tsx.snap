// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`success: vlag features renders correctly 1`] = `
<div
  className="vlag-features"
>
  <div
    className="vlag-features__item"
  >
    <div
      className="vlag-features__item__content"
    >
      <h2
        className="vlag-features__item__content__title"
      >
        Checkout Page
      </h2>
      <p
        className="vlag-features__item__content__description"
      >
        Lorem Ipsum is simply dummy text of the printing and typesetting industry. 
      </p>
      <p
        className="vlag-features__item__content__helper"
      >
        Will be available for beta testing end of 2022
      </p>
    </div>
    <div
      className="vlag-features__item__toggle"
    >
      <div
        className="vlag-switch"
      >
        <label
          className="vlag-switch__label"
        >
          <div
            className="vlag-switch__item  "
            onClick={[Function]}
          >
            <div
              className="vlag-switch__item__box"
            />
            <div
              className="vlag-switch__item__dots"
            />
          </div>
        </label>
      </div>
    </div>
  </div>
  <div
    className="vlag-features__item"
  >
    <div
      className="vlag-features__item__content"
    >
      <h2
        className="vlag-features__item__content__title"
      >
        Product Page
      </h2>
      <p
        className="vlag-features__item__content__description"
      >
        Lorem Ipsum is simply dummy text of the printing and typesetting industry. 
      </p>
    </div>
    <div
      className="vlag-features__item__toggle"
    >
      <div
        className="vlag-switch"
      >
        <label
          className="vlag-switch__label"
        >
          <div
            className="vlag-switch__item  "
            onClick={[Function]}
          >
            <div
              className="vlag-switch__item__box"
            />
            <div
              className="vlag-switch__item__dots"
            />
          </div>
        </label>
      </div>
    </div>
  </div>
</div>
`;
